### PR TITLE
Bech32m Support

### LINKED ISSUE #23: Taproot Witness Version
The [proposed Taproot BIP-341](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki) specifies version 1 witness scripts thus:

* Version number 1
* 32-byte witness program

### PR DESCRIPTION
**Pull request by clarkmoody**
Supports the new BIP-0350 Bech32m encoding for Bitcoin addresses.

Closes #23

### CODE REVIEW

**Review by sgeisler — COMMENTED** [review]
Looks mostly good.

Reading the code again I'm wondering if the design choices for the `WitnessProgram` type were right with keeping the serialized representation in there. But I guess if it's actually a problem someone will open a PR.

File: src/lib.rs [hunk 1]
```diff
@@ -218,6 +218,11 @@ impl FromStr for WitnessProgram {
             let program = Vec::from_base32(p5)?;
             (v[0], program)
         };
+        if (version.to_u8() == 0 && variant != Variant::Bech32)
```
[the remark this answers is not in the document]
Comment (clarkmoody) [reply src/lib.rs]: Good catch. I updated in fb193bc and also moved the encoding check after the rest of the validation. This matches the BIP-0350 order of operations and gives us error messages that match that order.

File: src/lib.rs [hunk 2]
```diff
@@ -225,10 +225,25 @@ impl FromStr for WitnessProgram {
             bech32: s.to_string(),
         };
         wp.validate()?;
+        if let Some(version_variant) = program_version_to_variant(version) {
+            if version_variant != variant {
+                return Err(Error::InvalidEncoding);
+            }
+        }
```
Comment (sgeisler) [inline src/lib.rs:232]: What happens if we parse an invalid version? As it is we don't check the variant at all. I think the else branch should return `Err(Error::InvalidScriptVersion)`.

Comment (clarkmoody) [reply src/lib.rs:232]: We call `wp.validate()?` before this final variant check, so we'll error out on script version and length prior to this point. I think we're following the [pseudo-code in BIP-0350](https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki#addresses-for-segregated-witness-outputs) pretty closely here, which performs the Bech32m check last.

Comment (sgeisler) [reply src/lib.rs:232]: Ah, I see, totally missed that!

**Review by sgeisler — APPROVED** [review]
ACK fb193bc19d17d5eb431d6fa5b17e10532fc218b1

**Pull request merged**